src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs

```diff
@@ -1073,52 +1084,122 @@ public async Task GoAwayFrame_AllPendingStreamsValid_RequestsSucceedAndConnectio
+                // Issue three requests, we want to make sure the specific task is related with specific stream
+                Task<HttpResponseMessage> sendTask1 = client.GetAsync("request1");
                 int streamId1 = await connection.ReadRequestHeaderAsync();
+
+                Task<HttpResponseMessage> sendTask2 = client.GetAsync("request2");
                 int streamId2 = await connection.ReadRequestHeaderAsync();
+                bool request2Retried = false;
+
+                Task<HttpResponseMessage> sendTask3 = client.GetAsync("request3");
                 int streamId3 = await connection.ReadRequestHeaderAsync();
+                bool request3Retried = false;
 
-                Assert.InRange(streamId1, int.MinValue, streamId2 - 1);
-                Assert.InRange(streamId2, int.MinValue, streamId3 - 1);
+                Assert.InRange(streamId1, 1, streamId2 - 1);
+                Assert.InRange(streamId2, streamId1 + 1, streamId3 - 1);
+                Assert.InRange(streamId3, streamId2 + 1, Int32.MaxValue);
 
                 // Send various partial responses
 
-                // First response: Don't send anything yet
+                // First response: Don't send anything yet, request should be retried on the new connection
 
                 // Second response: Send headers, no body yet
+                //   In case when headers were parsed we will fail this request
+                //   If headers were not parsed we will retry
+                //   Currently there is no way for server to tell if the server has parsed the headers.
+                //   Regardless of the case we want to make sure that something unexpected doesn't happen (i.e. hang)
+                //   Note: PING can tell that headers got received but does not guarantee them being parsed
+                //         (in most common case it will parse them but when client is experiencing heavy load it might not)
                 await connection.SendDefaultResponseHeadersAsync(streamId2);
 
                 // Third response: Send headers, partial body
+                //   As above, retry is not guaranteed
                 await connection.SendDefaultResponseHeadersAsync(streamId3);
                 await connection.SendResponseDataAsync(streamId3, new byte[5], endStream: false);
 
-                // Send a GOAWAY frame that indicates that we will abort all the requests.
-                var goAwayFrame = new GoAwayFrame(0, (int)ProtocolErrors.ENHANCE_YOUR_CALM, new byte[0], 0);
-                await connection.WriteFrameAsync(goAwayFrame);
+                // Ensure all sent frames are received by client
+                await connection.PingPong();
 
-                // We will not send any more frames, so send EOF now, and ensure the client handles this properly.
-                connection.ShutdownSend();
+                // Send a GOAWAY frame that indicates that we have not processed any of the requests
```

Comment:
I think you need to add a comment here, explaining that what the server is doing is technically illegal per RFC. That's why the nondeterminism occurs here with the second and third request.

Reply:
Also: do we have an issue tracking this nondeterminism, for the future?

I think the correct thing for us to do here is to fail these requests. The server has indicated it isn't going to process the requests any more, so we have to assume they are aborted. It's fine to punt this out past 3.0, but I still think we should have an issue to address this eventually.

Reply:
Though I also wonder if we can just fix the nondeterminism here in a simple way: if OnAbort is called with retry=true, and we're not in the ExpectingStatus state, then we must have already received something on this stream from the server, and we should throw a retryable exception. That would clean up the complications from nondeterminism here...

Reply:
I mean: we should NOT throw a retryable exception, will edit above for clarity

Reply:
@geoffkizer actually I think this is correct:
- server has not parsed any requests so lastStreamId is correctly 0
- for request 1, we never received any data so retry always occurs
- for request 2/3, we do know that some data was sent but we do not know if client has parsed it so we cannot tell if retry will or will not occur

the reason non-determinism occurs is because https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1554 works independently from frames processing and thus we cannot tell when the parsing over there happened.

In order for that to be deterministic we would need to make sure the frame processing loop waits for ReadResponseHeadersAsync to finish before proceeding with processing new frames.

Having said that, I do not believe we should be deterministic here becuase it would make the frame processing slower only for the sake of having deterministic tests - at least so far I couldn't justify making this change

Reply:
I agree that we shouldn't add any waits in ReadResponseHeadersAsync just to deal with this situation.

But I'm not clear that we need to. The processing of frames from the server is serialized. For request 2/3, we will have processed frames for these requests before we process the GOAWAY, and we will have transitioned the stream state past ExpectingStatus. So when we process the GOAWAY, we just need to check the stream state, and not create a retryable exception in this case.

Reply:
In other words, we don't need to change anything in the ReadResponseHeadersAsync path; the timing there doesn't matter, what matters is whether the abort exception has the retryable flag set or not.

Reply:
To clarify: we just need to check the stream state, and not set _canRetry = true in this case.